test/login_test/login_bloc_test.dart

```diff
@@ -0,0 +1,108 @@
+import 'package:bloc_test/bloc_test.dart';
+import 'package:flutter_test/flutter_test.dart';
+import 'package:mentorship_client/remote/requests/login.dart';
+import 'package:mentorship_client/remote/responses/auth_token.dart';
+import 'package:mentorship_client/screens/login/bloc/bloc.dart';
+
+import 'package:mockito/mockito.dart';
+import 'package:mentorship_client/remote/repositories/auth_repository.dart';
+import 'package:mentorship_client/auth/bloc.dart';
+
+class MockUserRepository extends Mock implements AuthRepository {}
+
+class MockAuthenticationBloc extends Mock implements AuthBloc {}
+
+void main() {
+  LoginBloc loginBloc;
+
+  MockUserRepository userRepository;
+  MockAuthenticationBloc authenticationBloc;
+  Login login;
+  AuthToken authToken;
+  setUp(() {
+    userRepository = MockUserRepository();
+    authenticationBloc = MockAuthenticationBloc();
+    loginBloc = LoginBloc(
+      userRepository,
+      authenticationBloc,
+    );
+    login = Login(
+      username: 'valid.username',
+      password: 'valid.password',
+    );
+
+    authToken = AuthToken(
+      'valid.token',
+      7,
+    );
+  });
+
+  tearDown(() {
+    loginBloc?.close();
+    authenticationBloc?.close();
+  });
+
+  test('initial state is correct', () {
+    expect(LoginInitial(), loginBloc.initialState);
+  });
+
+  test('close does not emit new states', () {
+    expectLater(
+      loginBloc,
+      emitsInOrder([LoginInitial(), emitsDone]),
+    );
+    loginBloc.close();
+  });
+
+  group('LoginButtonPressed', () {
+    blocTest(
+      'emits [LoginLoading, LoginSuccess] and token on success',
+      build: () async {
+        when(userRepository.login(
+          login,
+        )).thenAnswer((_) => Future.value(
+              authToken,
+            ));
+        return loginBloc;
+      },
+      act: (bloc) => bloc.add(
+        LoginButtonPressed(
+          login,
+        ),
+      ),
+      expect: [
+        LoginInProgress(),
+        LoginSuccess(),
+      ],
+      verify: (_) async {
+        verify(authenticationBloc.add(JustLoggedIn(
+          authToken.token,
+        ))).called(1);
+      },
+    );
+
+    blocTest(
+      'emits [LoginLoading, LoginFailure] on failure',
+      build: () async {
+        when(userRepository.login(
+          login,
+        )).thenThrow(Exception('login-error'));
```

Comment:
@bartekpacia regarding your comment, i removed Faliure because i wanted to throw 'login-errror' exception here and wasn't able to find a workaround. Do you have any suggestions?

Reply:
Unfortunately, I don't, I haven't done almost anything with this codebase for a few months and my knowledge about its state basically comes from reviewing PRs.
If your solution works fine, then use it. We can always refactor later.

Reply:
I still haven't really understood why is Failure used, 
> when something bizarre happens, Repository will return an exception and we need to catch it using a typical catch
(everything),

so if we have ```on Failure``` will we still be able to catch that? Dosen't it only catch Failure message strings from the backend?